### PR TITLE
Some Fixes in cuda Codegen regarding Subdomains

### PR DESCRIPTION
**Pull request by mroethlin**
## Technical Description

A few bug fixes related to subdomains:
* correct an off by one in horizontal size
* fixes number of threads vs stride confusion

Furthermore:
Small improvement in the vtk serialization code

### Note

Contains this [enum hack](https://github.com/MeteoSwiss-APN/dawn/pull/1108), tbd

### Testing

Can only be tested in integration

### Dependencies

This PR is independent

### CODE REVIEW

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

File: dawn/src/dawn/CodeGen/Cuda-ico/CudaIcoCodeGen.cpp [hunk 1]
```diff
@@ -315,28 +315,28 @@ void CudaIcoCodeGen::generateRunFun(
...
           return (iterSpace.has_value() ? "mesh_.Domain({::dawn::LocationType::Vertices," +
-                                              spaceMagicNumToEnum(iterSpace->upperBound()) + "," +
+                                              spaceMagicNumToEnum(iterSpace->upperLevel()) + "," +
                                               std::to_string(iterSpace->upperOffset()) + "})" +
                                               "- mesh_.Domain({::dawn::LocationType::Vertices," +
-                                              spaceMagicNumToEnum(iterSpace->lowerBound()) + "," +
-                                              std::to_string(iterSpace->lowerOffset()) + "})"
+                                              spaceMagicNumToEnum(iterSpace->lowerLevel()) + "," +
+                                              std::to_string(iterSpace->lowerOffset()) + "}) + 1"
                                         : "mesh_.NumVertices");
```
Comment (Stagno) [inline dawn/src/dawn/CodeGen/Cuda-ico/CudaIcoCodeGen.cpp:340]: Noticed that we have a lot of code repetition here: can factorize out common part and leave switch to map location type enum to `Cells`/`Edges`/`Vertices` string (we should have this function somewhere, maybe to the singular version of the string?).
```
          return (iterSpace.has_value() ? "mesh_.Domain({::dawn::LocationType::" + toStringPlural(loc) + "," +
                                              spaceMagicNumToEnum(iterSpace->upperBound()) + "," +
                                              spaceMagicNumToEnum(iterSpace->upperLevel()) + "," +
                                              std::to_string(iterSpace->upperOffset()) + "})" +
                                              "- mesh_.Domain({::dawn::LocationType::" + toStringPlural(loc) + "," +
                                              spaceMagicNumToEnum(iterSpace->lowerBound()) + "," +
                                              std::to_string(iterSpace->lowerOffset()) + "})"
                                              spaceMagicNumToEnum(iterSpace->lowerLevel()) + "," +
                                              std::to_string(iterSpace->lowerOffset()) + "}) + 1"
                                        : "mesh_.Num" + toStringPlural(loc) + ");
```

File: dawn/src/dawn/CodeGen/Cuda-ico/CudaIcoCodeGen.cpp [hunk 2]
```diff
@@ -347,33 +347,34 @@ void CudaIcoCodeGen::generateRunFun(
...
-                 spaceMagicNumToEnum(iterSpace.lowerBound()) + "," +
+                 spaceMagicNumToEnum(iterSpace.lowerLevel()) + "," +
                  std::to_string(iterSpace.lowerOffset()) + "})";
         case ast::LocationType::Vertices:
           return "mesh_.Domain({::dawn::LocationType::Vertices," +
-                 spaceMagicNumToEnum(iterSpace.lowerBound()) + "," +
+                 spaceMagicNumToEnum(iterSpace.lowerLevel()) + "," +
                  std::to_string(iterSpace.lowerOffset()) + "})";
         default:
           throw std::runtime_error("Invalid location type");
```
Comment (Stagno) [inline dawn/src/dawn/CodeGen/Cuda-ico/CudaIcoCodeGen.cpp:362]: Also here (see comment on code above)

**Review by Stagno — CHANGES_REQUESTED** [review]
Minor adjustment, then gtg.

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

File: dawn/src/dawn/CodeGen/Cuda-ico/CudaIcoCodeGen.cpp [hunk 3]
```diff
@@ -312,62 +312,25 @@ void CudaIcoCodeGen::generateRunFun(
...
-          throw std::runtime_error("Invalid location type");
-        }
       };
 
       auto domain = stage->getUnstructuredIterationSpace();
       std::string hSizeString = "hsize" + std::to_string(stage->getStageID());
       std::string numElString = "mesh_." + locToDenseSizeStringGpuMesh(*stage->getLocationType(), codeGenOptions.UnstrPadding);
       std::string hOffsetString = "hoffset" + std::to_string(stage->getStageID());
-      runFun.addStatement("int " + hSizeString + " = " +
+      runFun.addStatement("int " + hSizeString + " = " g
```
Comment (Stagno) [inline dawn/src/dawn/CodeGen/Cuda-ico/CudaIcoCodeGen.cpp:333]:
```suggestion
      runFun.addStatement("int " + hSizeString + " = " +
```

**Review by Stagno — CHANGES_REQUESTED** [review]
Spotted a typo

**Review by Stagno — APPROVED** [review]
Approved after typo fix

**Comment by mroethlin** [comment]
launch jenkins

**Pull request merged**